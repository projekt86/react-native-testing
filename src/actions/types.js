// @flow

export type Action =
  | { type: "LOGGING" }
  | { type: "LOGGED_IN", data: { id: string, name: string } }
  | { type: "LOGGING_ERROR", error: string }
  | { type: "LOGGED_OUT" };

export type Dispatch = (
  action: Action | ThunkAction | PromiseAction | Array<Action>
) => any;
export type GetState = () => Object;
export type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;
export type PromiseAction = Promise<Action>;
