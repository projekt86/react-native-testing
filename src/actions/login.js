// @flow
import axios from 'axios';
import { ThunkAction, Action } from './types';


export const login = (email: string, password: string): Promise => {
  return (dispatch) => {
    dispatch({ type: 'LOGGING' });

    return axios
      .post('https://us-central1-test-21181.cloudfunctions.net/api/login', { email, password })
      .then(response => dispatch(loginSuccess(response.data)))
      .catch(error => dispatch(loginError('Invalid email or password')));
  }
}

const loginSuccess = (data: {id: string, name: string}): Action => {
    return {
      type: 'LOGGED_IN',
      data: { id: data.id, name: data.name }
    };
};

const loginError = (error: string): Action => {
  return {
    type: 'LOGGING_ERROR',
    error
  };
};

export const logout = (): Action => {
  return {
    type: 'LOGGED_OUT'
  };
};
