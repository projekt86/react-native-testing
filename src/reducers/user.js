/**
* @flow
*/


import type { Action } from "../actions/types";

export type State = {
  isLoggedIn: boolean,
  id?: string,
  name?: string,
  isLoading: boolean,
  error?: string
};

export const initialState = {
  isLoggedIn: false,
  id: null,
  name: null,
  isLoading: false,
  error: null,
};

function user(state: State = initialState, action: Action): State {
  if (action.type === "LOGGED_IN") {
    let { id, name } = action.data;
    return {
      isLoggedIn: true,
      id,
      name,
      isLoading: false,
      error: null,
    };
  }
  if (action.type === "LOGGED_OUT") {
    return { ...initialState };
  }
  if (action.type === 'LOGGING') {
    return { ...initialState, isLoading: true };
  }
  if (action.type === 'LOGGING_ERROR') {
    return { ...initialState, error: action.error };
  }
  return state;
}

module.exports = user;
