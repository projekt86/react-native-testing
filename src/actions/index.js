import * as loginActions from "./login";

module.exports = {
  ...loginActions
};
