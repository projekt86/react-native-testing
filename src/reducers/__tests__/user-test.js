import user from '../user';
import { initialState } from '../user';
import { login } from '../../actions/login';

it('returns the same state on unhandled action ', () => {
  expect(user(initialState, { type: '_NULL'})).toMatchSnapshot();
});

it('handles LOGGING action', () => {
  expect(user(initialState, { type: 'LOGGING'})).toEqual({
    id: null, name: null, isLoggedIn: false, isLoading: true, error: null
  });
});

it('handles LOGGED_IN action', () => {
  expect(user(initialState, { type: 'LOGGED_IN', data: { id: 'A', name: 'B'}})).toEqual({
    id: 'A', name: 'B', isLoggedIn: true, isLoading: false, error: null
  });
});

it('handles LOGGING_ERROR action', () => {
  expect(user(initialState, { type: 'LOGGING_ERROR', error: 'A'})).toEqual({
    id: null, name: null, isLoggedIn: false, isLoading: false, error: 'A'
  });
});

it('handles LOGGED_OUT action', () => {
  expect(user(initialState, { type: 'LOGGED_OUT'})).toMatchSnapshot();
});
