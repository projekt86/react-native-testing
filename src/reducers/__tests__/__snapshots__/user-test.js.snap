// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles LOGGED_OUT action 1`] = `
Object {
  "error": null,
  "id": null,
  "isLoading": false,
  "isLoggedIn": false,
  "name": null,
}
`;

exports[`returns the same state on unhandled action  1`] = `
Object {
  "error": null,
  "id": null,
  "isLoading": false,
  "isLoggedIn": false,
  "name": null,
}
`;
